cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/sourcecontrol/InMemorySourceControlOperationRunner.java

```diff
@@ -80,37 +81,59 @@ public PushAppsResponse push(List<ApplicationDetail> appsToPush, CommitMeta comm
 
   private void modifyApplicationConfigAndPush(List<ApplicationDetail> appsToPush, CommitMeta commitDetails)
     throws PushFailureException {
+    List<Path> modifiedPaths = new ArrayList<Path>();
     try {
       // Create base dir if not exist.
       // This already handles cases where one of the path prefix is not a directory
       Files.createDirectories(repositoryManager.getBasePath());
 
       for (ApplicationDetail appDetail : appsToPush) {
         String appName = appDetail.getName();
-        Path configFilePath = getApplicationConfigPath(appName);
+        Path configFilePath = generateAndValidateApplicationConfigPath(appName);
         byte[] configData = GSON.toJson(appDetail).getBytes(StandardCharsets.UTF_8);
+
         // Opens the file for writing, creating the file if it doesn't exist,
         // or truncating an existing regular-file to a size of 0
         // Symlinks will be followed but if the path points to a directory it will fail
         Files.write(configFilePath, configData);
+        modifiedPaths.add(configFilePath);
 
         LOG.debug("Written application configs for {} in file {}", appName, configFilePath);
       }
-    } catch (IOException e) {
+    } catch (Exception e) {
       LOG.debug("Failed to write configs: %s", e.getCause());
       throw new PushFailureException("Failed to write configs", e);
     }
 
     try {
-      repositoryManager.commitAndPush(commitDetails);
+      repositoryManager.commitAndPush(commitDetails, modifiedPaths);
       LOG.info("Pushed application configs for : {}", appsToPush.stream().map(ApplicationDetail::getName));
     } catch (Exception e) {
       LOG.debug("Failed to push to git", e.getCause());
       throw new PushFailureException("Failed to push to git", e);
     }
   }
-  private Path getApplicationConfigPath(String applicationName) {
-    return repositoryManager.getBasePath().resolve(String.format("%s.json", applicationName));
+
+  /**
+   * Generates the config file path in git based on application name.
+   * It also validates if the resolved path ( considering symlinks) exists outside git repository
+   * or not.
+   *
+   * @param applicationName
+   * @return A valid application config file path
+   */
+  private Path generateAndValidateApplicationConfigPath(String applicationName) throws InvalidPathInSourceControl,
+    IOException {
+    Path filePath = repositoryManager.getBasePath().resolve(String.format("%s.json", applicationName));
+    if (Files.exists(filePath)) {
+      Path resolvedPath = filePath.toRealPath();
+      if (!resolvedPath.startsWith(repositoryManager.getRepositoryRoot())) {
```

Comment:
@samdgupi This does not work for macOS: https://apple.stackexchange.com/questions/1043/why-is-tmp-a-symlink-to-private-tmp

Instead I made the changes to compare the `RealPath` of both and return a relative path string